content/blog/2019/06/2019-06-10-use-casc-for-exws.adoc

```diff
@@ -0,0 +1,129 @@
+---
+layout: post
+title: "Use Terraform with External Workspace Manager to Support EFS"
+tags:
+- jenkins
+- gsoc
+author: yufei_zhang
+description: "A brief example of using Terraform with EFS" # optional
+opengraph:
+---
+
+=== Why you want to use External Workspace Manager with EFS
+What is Terraform : A CasC software, like CloudFormation and Ansible designed for managing cloud resources. It can free you from managing cloud resource by console, and can be version controlled using git as it is configuration-as-code. Currently it supports AWS and Azure best. AWS CloudFormation is for AWS only.
+What is extenal-workspace-manager : It is a pipeline plugin let you custom your workspace path, you can mount shared disk to multiple nodes to share workspaces.
+What is EFS : Amazon Elastic File System, an Network File System, can be shared among a lot of EC2 instances.
+
+Who should read this blog : build Jenkins based on cloud, want to have good version control over all resources including infrastructure; You want to share workspace among agents (slaves), this might because you have a upstream job creating a large file and don't want to copy it into downstream job's workspace because it's time-consuming; You're afraid of running out of workspaces, and wants an extensible disk.
+
+Who should not read this blog : 
+. Skilled AWS architecture already using CasC tools
+. Ops already using CasC and know how to provision EFS to a Jenkins node using user data.
+. Running Jenkins in local physical machine, and do not build Jenkins based on Clouds.
+
+What you'll want to do, is to learn external-workspace-manager directly to do similiar things, see blog link : link:https://github.com/jenkinsci/external-workspace-manager-plugin [External Workspace Manager]
+
+This tutorial use AWS as example, but other clouds like Google Cloud, Microsoft Azure are also supported. (same mechanism, different syntax), and here we go.
+
+=== Define a master Jenkins EC2
+
+Terraform use its own DSL, tutorial can be found : , but you don't need to worry about the syntaxes. You only need to know the concept of CasC, then you can use any CasC with any cloud providers or even other services.You create a ```jenkins.tf``` file, and then write such code :
+
+```
+provider "aws" {
+  profile    = "default"
+  region     = "us-east-1"
+}
+
+```
+This setup the AWS account you want to use. See more detail on : link:https://www.terraform.io/intro/index.html [introduction on terraform]
+```
+resource "aws_instance" "jenkins_master" {
+  ami           = "ami-174325867235"
+  instance_type = "t2.micro"
+  *user_data*    = "${data.template_file.jenkins_start_script}"
+}
+```
+This defines a ec2 instance, and use a template we will define later as the user data. User data is a script will be executed when instance launched. We will use this user data to install dependencies, Jenkins, mount EFS and many other things.
```

Comment:
"User data is a script will be ..." -> "User data is a script to be executed when the instance is launched".